creme/stream/datasets/agrawal.py

```diff
@@ -0,0 +1,316 @@
+                hyears = 1 + self._random_state.randint(30)
+                loan = self._random_state.rand() * 500000
+                group = self._classification_functions[self.classification_function](salary,
+                                                                                     commission,
+                                                                                     age, elevel,
+                                                                                     car,
+                                                                                     zipcode,
+                                                                                     hvalue,
+                                                                                     hyears, loan)
+                if not self.balance_classes:
+                    desired_class_found = True
+                else:
+                    if (self._next_class_should_be_zero and (group == 0)) or \
+                            ((not self._next_class_should_be_zero) and (group == 1)):
+                        desired_class_found = True
+                        self._next_class_should_be_zero = not self._next_class_should_be_zero
+
+            if self.perturbation > 0.0:
+                salary = self._perturb_value(salary, 20000, 150000)
+                if commission > 0:
+                    commission = self._perturb_value(commission, 10000, 75000)
+                age = np.round(self._perturb_value(age, 20, 80))
+                hvalue = self._perturb_value(hvalue, (9 - zipcode) * 100000, 0, 135000)
+                hyears = np.round(self._perturb_value(hyears, 1, 30))
+                loan = self._perturb_value(loan, 0, 500000)
+
+            x = dict()
+            for key in self.feature_names:
+                x[key] = eval(key)
+            y = group
+
+            self.current_sample_x = x
+            self.current_sample_y = y
+
+            yield x, y
+
+    def _perturb_value(self, val, val_min, val_max, val_range=None):
+        if val_range is None:
+            val_range = val_max - val_min
+        val += val_range * (2 * (self._random_state.rand() - 0.5)) * self.perturbation
+        if val < val_min:
+            val = val_min
+        elif val > val_max:
+            val = val_max
+        return val
+
+    def generate_drift(self):
+        """
+        Generate drift by switching the classification function randomly.
+
+        """
+        new_function = self._random_state.randint(10)
+        while new_function == self.classification_function:
+            new_function = self._random_state.randint(10)
+        self.classification_function = new_function
+
+    @staticmethod
+    def _classification_function_0(salary, commission, age, elevel, car, zipcode, hvalue,
+                                   hyears, loan):
+        return 0 if ((age < 40) or (60 <= age)) else 1
```

Comment:
This could be rewritten as `int(age < 40 or 60 <= age)`. The same goes for the rest of the code. It's not very important though. I don't mind cleaning things up once this is merged. I actually enjoy doing that kind of work :)

Reply:
Agree. This is a good time to revisit the code and consider cosmetic changes that can improve the code base.